File: code/java/剑指offer33.java
class Solution {
    public boolean judege(int[] midorder,int[] postorder,int l_m,int r_m,int l_p,int r_p){
        if(l_m>r_m) return true;
        if(l_m==r_m) return midorder[l_m]==postorder[l_p];
        int root = postorder[r_p];
        int i = l_m;
        while(i<=r_m){
            if(midorder[i]==root) break;
            i+=1;
        }
        if(i>r_m) return false;
        boolean left = judege(midorder,postorder,l_m,i-1,l_p,l_p+(i-l_m)-1);
        boolean right = judege(midorder,postorder,i+1,r_m,l_p+(i-l_m),r_p-1);
        return left&right;
    }

    public boolean verifyPostorder(int[] postorder) {
        int[] midorder = new int[postorder.length];
        for(int i=0;i<midorder.length;i++) midorder[i]=postorder[i];
        Arrays.sort(midorder);
        return judege(midorder,postorder,0,midorder.length-1,0,midorder.length-1);
    }
}

File: code/python/剑指offer33.py
class Solution:
    def verifyPostorder(self, postorder: List[int]) -> bool:
        midorder = sorted(postorder)
        print(midorder)

        def judege(mid_l, mid_r, post_l, post_r):
            if mid_l > mid_r:
                return True
            if mid_l == mid_r:
                return midorder[mid_l] == postorder[post_l]

            root = postorder[post_r]
            i = mid_l
            while i <= mid_r:
                if midorder[i] == root:
                    break
                i += 1
            if i > mid_r:
                return False
            left = judege(mid_l, i - 1, post_l, post_l + (i - mid_l) - 1)
            right = judege(i + 1, mid_r, post_l + (i - mid_l), post_r - 1)
            # print(mid_l,i-1,post_l,post_l+(i-mid_l)-1)
            # print(i+1,mid_r,post_l+(i-mid_l),post_r-1)
            # print(i,mid_l,mid_r,post_l,post_r,left,right)
            return left and right

        return judege(0, len(postorder) - 1, 0, len(postorder) - 1)